algorithms/cuda/src/clusterization/component_connection_kernels.cu

```diff
@@ -0,0 +1,185 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ * 
+ * (c) 2021 CERN for the benefit of the ACTS project
+ * 
+ * Mozilla Public License Version 2.0
+ */
+
+#include "clusterization/component_connection_kernels.cuh"
+#include "../../cuda/src/utils/cuda_error_handling.hpp" 
+#include <iostream>
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+namespace traccc{
+
+__device__
+unsigned int find_root(const vecmem::device_vector< unsigned int >& L, unsigned int e){
+  unsigned int r = e;
+  while (L[r] != r){
+    r = L[r];
+  }
+  return r;
+}
+
+__device__
+unsigned int make_union(vecmem::device_vector< unsigned int >& L, unsigned int e1, unsigned int e2){
+  int e;
+  if (e1 < e2){
+    e = e1;
+    L[e2] = e;
+  } else {
+    e = e2;
+    L[e1] = e;
+  }
+  return e;
+}
+
+__device__
+bool is_adjacent(cell a, cell b){
+  return (a.channel0 - b.channel0)*(a.channel0 - b.channel0) <= 1 
+    and (a.channel1 - b.channel1)*(a.channel1 - b.channel1) <= 1; 
+}
+
+__device__
+bool is_far_enough(cell a, cell b){
+  return (a.channel1 - b.channel1) > 1; 
+}
+
+__device__
+unsigned int sparse_ccl(const vecmem::device_vector< cell > cells,
+			vecmem::device_vector< unsigned int >& L){
+
+  unsigned int start_j = 0;
+  for (unsigned int i=0; i<cells.size(); ++i){
+    L[i] = i;
+    int ai = i;
+    if (i > 0){
+      for (unsigned int j = start_j; j < i; ++j){
+	if (is_adjacent(cells[i], cells[j])){
+	  ai = make_union(L, ai, find_root(L, j));
+	} else if (is_far_enough(cells[i], cells[j])){
+	  ++start_j;
+	}
+      }
+    }    
+  }
+
+  // second scan: transitive closure
+  unsigned int labels = 0;
+  for (unsigned int i = 0; i < cells.size(); ++i){
+    unsigned int l = 0;
+    if (L[i] == i){
+      ++labels;
+      l = labels; 
+    } else {
+      l = L[L[i]];
+    }
+    L[i] = l;
+  }
+  
+  return labels;
+}
+
+
+__global__
+void sparse_ccl_kernel(vecmem::data::jagged_vector_view< cell > _cell_per_event,
+		     vecmem::data::jagged_vector_view< unsigned int > _label_per_event,
+		     vecmem::data::vector_view<unsigned int> _num_labels);
+
+__global__
+void sp_formation_kernel(vecmem::data::jagged_vector_view< cell > _cell_per_event,
+		       vecmem::data::jagged_vector_view< unsigned int> _label_per_event,
+		       vecmem::data::vector_view<unsigned int> _num_labels,
+		       vecmem::data::jagged_vector_view< measurement > _ms_per_event,
+		       vecmem::data::jagged_vector_view< spacepoint > _sp_per_event);
+
+  void sparse_ccl_cuda(
+		  const vecmem::data::jagged_vector_view< cell >& cell_per_event,
+		  vecmem::data::jagged_vector_view< unsigned int>& label_per_event,
+		  vecmem::data::vector_view<unsigned int> num_labels){    
+    sparse_ccl_kernel<<< 1, cell_per_event.m_size >>>(cell_per_event,
+						    label_per_event,
+						    num_labels);        
+    VECMEM_CUDA_ERROR_CHECK(cudaGetLastError());
+    VECMEM_CUDA_ERROR_CHECK(cudaDeviceSynchronize());
```

Comment:
Never mind... Apparently "WIP:" is meaningful here as well. It's just quite confusing that there is a separate mechanism in place as well for declaring a PR as a draft... 😕

Reply:
Yeah, one of them is through a Bot that's added to the org and the other is GH native. We use the bot on the ACTS main repo to prevent merging WIP PRs. I think I can disable the Bot for this repo if you prefer this.